src/Integration/Aws/DynamoDbSession/src/SessionHandler.php

```diff
@@ -0,0 +1,206 @@
+                ],
+            ],
+            'KeySchema' => [
+                [
+                    'AttributeName' => $this->config['hash_key'],
+                    'KeyType' => KeyType::HASH,
+                ],
+            ],
+        ]);
+
+        do {
+            sleep($wait);
+
+            $table = $this->client->describeTable(['TableName' => $this->config['table_name']])->getTable();
+
+            if (!$table) {
+                throw new \RuntimeException(sprintf('Could not create table %s', $this->config['table_name']));
+            }
+        } while (TableStatus::ACTIVE !== $table->getTableStatus());
+
+        $this->client->updateTimeToLive([
+            'TableName' => $this->config['table_name'],
+            'TimeToLiveSpecification' => [
+                'Enabled' => true,
+                'AttributeName' => $this->config['session_lifetime_attribute'],
+            ],
+        ]);
+    }
+
+    public function close()
+    {
+        $id = \session_id();
+
+        // Make sure the expiration time is updated, even if the write did not occur
+        if ($this->sessionId !== $id || !$this->sessionWritten) {
+            $this->sessionWritten = $this->doWrite($id, false);
+        }
+
+        return $this->sessionWritten;
+    }
+
+    public function destroy($session_id)
+    {
+        $this->sessionId = $session_id;
+
+        $this->client->deleteItem([
+            'TableName' => $this->config['table_name'],
+            'Key' => $this->formatKey($this->formatId($session_id)),
+        ]);
+
+        return $this->sessionWritten = true;
+    }
+
+    public function gc($maxlifetime)
+    {
+        // DynamoDB takes care of garbage collection
+        return true;
+    }
+
+    public function open($save_path, $name)
+    {
+        $this->sessionName = $name;
+
+        return true;
+    }
+
+    public function read($session_id)
+    {
+        $this->sessionId = $session_id;
+
+        $this->dataRead = '';
+
+        $attributes = $this->client->getItem([
+            'TableName' => $this->config['table_name'],
+            'Key' => $this->formatKey($this->formatId($session_id)),
+            'ConsistentRead' => $this->config['consistent_read'] ?? true,
+        ])->getItem();
+
+        // Return the data if it is not expired. If it is expired, remove it
+        if (isset($attributes[$this->config['session_lifetime_attribute']]) && isset($attributes[$this->config['data_attribute']])) {
+            $this->dataRead = $attributes[$this->config['data_attribute']]->getS() ?? '';
+            if ($attributes[$this->config['session_lifetime_attribute']]->getN() <= time()) {
+                $this->dataRead = '';
+                $this->destroy($session_id);
+            }
+        }
+
+        return $this->dataRead;
+    }
+
+    public function write($session_id, $session_data)
+    {
+        $changed = $session_id !== $this->sessionId
+            || $session_data !== $this->dataRead;
```

Comment:
Make this one line so it is easier to read. 


Also, is this correct? If the value didnt change, then we dont update the `$expires` key.

Reply:
I had slightly rewritten it in https://github.com/async-aws/aws/pull/617/commits/cd572be05a57a587bd90f07fd94cd981910053b5 to make it clearer, but apparently it's not clear enough 😀 

The `$updateData` only affects the data attribute, the expires attribute is always updated

this has probably been introduced to prevent race conditions
(if two requests with the same session enter at the "same" time; only the first one updates the session, both requests will update the expiration time but the session data will be written only by the request that changed it, since the expiration time is very similar in both requests we don't care about race condition here)